src/server/game/Entities/Unit/Unit.cpp

```diff
@@ -12398,6 +12376,35 @@ void Unit::UpdateSpeed(UnitMoveType mtype, bool forced)
             speed = min_speed;
     }
 
+    if (GetTypeId() == TYPEID_UNIT)
+    {
+        switch (mtype)
+        {
+            case MOVE_RUN:
+                // Set creature speed rate
+                speed *= ToCreature()->GetCreatureTemplate()->speed_run;
+            default:
+                break;
+        }
+        
+        if (Unit* owner = GetCharmerOrOwner())
+        {
+            if ((IsGuardian() || IsPet()) && ToCreature()->HasUnitState(UNIT_STATE_FOLLOW) && !ToCreature()->IsInCombat())
+            {
+                float distance = GetDistance(owner);
+                float owner_speed = owner->GetSpeedRate(mtype);
+
+                // Sync speed with owner when near or slower
+                if (distance < PET_FOLLOW_DIST + sWorld->getRate(RATE_TARGET_POS_RECALCULATION_RANGE) || speed < owner_speed)
+                    speed = owner_speed;
+
+                // Pet speed decreases when near to help prevent stop-and-go movement
+                // and increases when away to help prevent falling behind.
+                speed *= ((distance < 3 ? distance : 3) + 8) / 10;
```

Comment:
Updated in [15e39ce](https://github.com/gashole/TrinityCore/commit/15e39ce57449a39cbe83f8b8b4fb39a9b7cd5bfd)